### DISCUSSION TITLE
Failed to compile ROSCO 2.4.1 using ifort

### DISCUSSION THREAD

**Discussion by yisehakabreham** [NatLabRockies/ROSCO#111] [Q&A] [answered]
Dear ROSCO community,

I have tried but failed to compile a DLL file using Intel Fortran Compiler (ifort). It works fine with gfortran (MinGW64), but I need to test a ROSCO controller compiled with ifort on a Commercial Wind Turbine program.
I tried on both the Intel command window and Visual Studio 2019. 

The error log is listed below,

Build started...
1>------ Build started: Project: LibDISCON (IFORT), Configuration: Debug x64 ------
1>Compiling with Intel® Fortran Compiler Classic 2021.5.0 [Intel(R) 64]...
1>Functions.f90
1>C:\Users\Yisehak\ROSCO\ROSCO\src\Functions.f90(504): error #6633: The type of the actual argument differs from the type of the dummy argument.   [CPFUNCTION]
1>compilation aborted for C:\Users\Yisehak\ROSCO\ROSCO\src\Functions.f90 (code 1)
1>ReadSetParameters.f90
1>C:\Users\Yisehak\ROSCO\ROSCO\src\ReadSetParameters.f90(19): error #7002: Error in opening the compiled module file.  Check INCLUDE paths.   [FUNCTIONS]
1>C:\Users\Yisehak\ROSCO\ROSCO\src\ReadSetParameters.f90(635): error #6404: This name does not have a type, and must have an explicit type.   [NONDECREASING]
1>C:\Users\Yisehak\ROSCO\ROSCO\src\ReadSetParameters.f90(937): error #6404: This name does not have a type, and must have an explicit type.   [INT2LSTR]
1>C:\Users\Yisehak\ROSCO\ROSCO\src\ReadSetParameters.f90(937): error #6362: The data types of the argument(s) are invalid.   [TRIM]
1>C:\Users\Yisehak\ROSCO\ROSCO\src\ReadSetParameters.f90(1001): error #6404: This name does not have a type, and must have an explicit type.   [INT2LSTR]
1>C:\Users\Yisehak\ROSCO\ROSCO\src\ReadSetParameters.f90(1001): error #6362: The data types of the argument(s) are invalid.   [TRIM]
1>C:\Users\Yisehak\ROSCO\ROSCO\src\ReadSetParameters.f90(1065): error #6404: This name does not have a type, and must have an explicit type.   [INT2LSTR]
1>C:\Users\Yisehak\ROSCO\ROSCO\src\ReadSetParameters.f90(1065): error #6362: The data types of the argument(s) are invalid.   [TRIM]
1>C:\Users\Yisehak\ROSCO\ROSCO\src\ReadSetParameters.f90(1130): error #6404: This name does not have a type, and must have an explicit type.   [INT2LSTR]
1>C:\Users\Yisehak\ROSCO\ROSCO\src\ReadSetParameters.f90(1130): error #6362: The data types of the argument(s) are invalid.   [TRIM]
1>C:\Users\Yisehak\ROSCO\ROSCO\src\ReadSetParameters.f90(1170): error #6362: The data types of the argument(s) are invalid.   [TRIM]
1>C:\Users\Yisehak\ROSCO\ROSCO\src\ReadSetParameters.f90(1257): error #6404: This name does not have a type, and must have an explicit type.   [INT2LSTR]
1>C:\Users\Yisehak\ROSCO\ROSCO\src\ReadSetParameters.f90(1257): error #6362: The data types of the argument(s) are invalid.   [TRIM]
1>C:\Users\Yisehak\ROSCO\ROSCO\src\ReadSetParameters.f90(1297): error #6362: The data types of the argument(s) are invalid.   [TRIM]
1>C:\Users\Yisehak\ROSCO\ROSCO\src\ReadSetParameters.f90(1374): error #6404: This name does not have a type, and must have an explicit type.   [INT2LSTR]
1>C:\Users\Yisehak\ROSCO\ROSCO\src\ReadSetParameters.f90(1374): error #6362: The data types of the argument(s) are invalid.   [TRIM]
1>C:\Users\Yisehak\ROSCO\ROSCO\src\ReadSetParameters.f90(1385): error #6362: The data types of the argument(s) are invalid.   [TRIM]
1>compilation aborted for C:\Users\Yisehak\ROSCO\ROSCO\src\ReadSetParameters.f90 (code 1)
1>ControllerBlocks.f90
1>C:\Users\Yisehak\ROSCO\ROSCO\src\ControllerBlocks.f90(20): error #7002: Error in opening the compiled module file.  Check INCLUDE paths.   [FUNCTIONS]
1>C:\Users\Yisehak\ROSCO\ROSCO\src\ControllerBlocks.f90(57): error #6404: This name does not have a type, and must have an explicit type.   [SATURATE]
1>C:\Users\Yisehak\ROSCO\ROSCO\src\ControllerBlocks.f90(216): error #6404: This name does not have a type, and must have an explicit type.   [AERODYNTORQUE]
1>C:\Users\Yisehak\ROSCO\ROSCO\src\ControllerBlocks.f90(246): error #6404: This name does not have a type, and must have an explicit type.   [INTERP1D]
1>C:\Users\Yisehak\ROSCO\ROSCO\src\ControllerBlocks.f90(250): error #6404: This name does not have a type, and must have an explicit type.   [INTERP2D]
1>C:\Users\Yisehak\ROSCO\ROSCO\src\ControllerBlocks.f90(279): error #6404: This name does not have a type, and must have an explicit type.   [IDENTITY]
1>C:\Users\Yisehak\ROSCO\ROSCO\src\ControllerBlocks.f90(361): error #6404: This name does not have a type, and must have an explicit type.   [INTERP1D]
1>compilation aborted for C:\Users\Yisehak\ROSCO\ROSCO\src\ControllerBlocks.f90 (code 1)
1>Controllers.f90
1>C:\Users\Yisehak\ROSCO\ROSCO\src\Controllers.f90(18): error #7002: Error in opening the compiled module file.  Check INCLUDE paths.   [FUNCTIONS]
1>C:\Users\Yisehak\ROSCO\ROSCO\src\Controllers.f90(20): error #7002: Error in opening the compiled module file.  Check INCLUDE paths.   [CONTROLLERBLOCKS]
1>C:\Users\Yisehak\ROSCO\ROSCO\src\Controllers.f90(55): error #6404: This name does not have a type, and must have an explicit type.   [INTERP1D]
1>C:\Users\Yisehak\ROSCO\ROSCO\src\Controllers.f90(62): error #6404: This name does not have a type, and must have an explicit type.   [PICONTROLLER]
1>C:\Users\Yisehak\ROSCO\ROSCO\src\Controllers.f90(83): error #6404: This name does not have a type, and must have an explicit type.   [PITCHSATURATION]
1>C:\Users\Yisehak\ROSCO\ROSCO\src\Controllers.f90(98): error #6404: This name does not have a type, and must have an explicit type.   [SHUTDOWN]
1>C:\Users\Yisehak\ROSCO\ROSCO\src\Controllers.f90(102): error #6404: This name does not have a type, and must have an explicit type.   [SATURATE]
1>C:\Users\Yisehak\ROSCO\ROSCO\src\Controllers.f90(103): error #6404: This name does not have a type, and must have an explicit type.   [RATELIMIT]
1>C:\Users\Yisehak\ROSCO\ROSCO\src\Controllers.f90(162): error #6404: This name does not have a type, and must have an explicit type.   [PICONTROLLER]
1>C:\Users\Yisehak\ROSCO\ROSCO\src\Controllers.f90(163): error #6404: This name does not have a type, and must have an explicit type.   [SATURATE]
1>C:\Users\Yisehak\ROSCO\ROSCO\src\Controllers.f90(193): error #6404: This name does not have a type, and must have an explicit type.   [RATELIMIT]
1>C:\Users\Yisehak\ROSCO\ROSCO\src\Controllers.f90(279): error #6404: This name does not have a type, and must have an explicit type.   [PICONTROLLER]
1>C:\Users\Yisehak\ROSCO\ROSCO\src\Controllers.f90(290): error #6404: This name does not have a type, and must have an explicit type.   [SATURATE]
1>C:\Users\Yisehak\ROSCO\ROSCO\src\Controllers.f90(341): error #6404: This name does not have a type, and must have an explicit type.   [PICONTROLLER]
1>C:\Users\Yisehak\ROSCO\ROSCO\src\Controllers.f90(366): error #6404: This name does not have a type, and must have an explicit type.   [PICONTROLLER]
1>C:\Users\Yisehak\ROSCO\ROSCO\src\Controllers.f90(411): error #6404: This name does not have a type, and must have an explicit type.   [PIICONTROLLER]
1>C:\Users\Yisehak\ROSCO\ROSCO\src\Controllers.f90(433): error #6404: This name does not have a type, and must have an explicit type.   [SATURATE]
1>compilation aborted for C:\Users\Yisehak\ROSCO\ROSCO\src\Controllers.f90 (code 1)
1>DISCON.F90
1>C:\Users\Yisehak\ROSCO\ROSCO\src\DISCON.F90(31): warning #5117: Bad # preprocessor line
1>C:\Users\Yisehak\ROSCO\ROSCO\src\DISCON.F90(34): warning #5117: Bad # preprocessor line
1>C:\Users\Yisehak\ROSCO\ROSCO\src\DISCON.F90(22): error #7002: Error in opening the compiled module file.  Check INCLUDE paths.   [READSETPARAMETERS]
1>C:\Users\Yisehak\ROSCO\ROSCO\src\DISCON.F90(23): error #7002: Error in opening the compiled module file.  Check INCLUDE paths.   [CONTROLLERBLOCKS]
1>C:\Users\Yisehak\ROSCO\ROSCO\src\DISCON.F90(24): error #7002: Error in opening the compiled module file.  Check INCLUDE paths.   [CONTROLLERS]
1>C:\Users\Yisehak\ROSCO\ROSCO\src\DISCON.F90(27): error #7002: Error in opening the compiled module file.  Check INCLUDE paths.   [FUNCTIONS]
1>compilation aborted for C:\Users\Yisehak\ROSCO\ROSCO\src\DISCON.F90 (code 1)
1>
1>Build log written to  "file://C:\Users\Yisehak\ROSCO\ROSCO\src\x64\Debug\BuildLog.htm"
1>LibDISCON - 51 error(s), 2 warning(s)
========== Build: 0 succeeded, 1 failed, 0 up-to-date, 0 skipped ==========

Can anyone help me with fixing or describing the problem?
Thank you in advance.

**Comment by ghylander** [comment, marked as answer]
Hi, 
the error you are getting is well-known, but it shouldn't be happening.
the actual error is that ifort expects some variables to be explicitly defined. The first time this happens is in Functions.f90 CPFUNCTION, then repeats.

This, however, should be taken care of by the CMakeLists.txt config for the ifort compiler

may I ask how are you attempting the compilation?

This is my setup and workflow, it works for my team and I:
CMake (in PATH)
ifort from Intel oneAPI (from HPC toolkit v2022.1.2-116)
Visual Studio (2017, 2019 & 2022)

Open terminal in (or change dir to) ~/ROSCO/ROSCO/build/
Execute, for the version of VS you want to generate the files for:
- VS2017: `cmake .. -G "Visual Studio 15 2017 Win64"`
- VS2019: `cmake .. -G "Visual Studio 16 2019" -A x64`
- VS2022: `cmake .. -G "Visual Studio 17 2022" -A x64`

Open the ROSCO.sln file with the corresponding VS version
Change build configuration (if required)
Compile solution

If this doesn't work for you, would you mind copying the contents of your CMakeLists.txt file and posting it here as a reply to this comment? Simply to make sure you are using the latest version of that file

**Reply by yisehakabreham** [reply]
@ghylander ,
thank you for providing the solution.
I used the CMakeLists.txt file you modified for Intel Fortran Compiler in the Pull repository and followed the above steps to successfully compile the DLL.

The dynamic link is now working as intended. Thank you again.

**Reply by ghylander** [reply]
glad to hear you got it working
